### DISCUSSION TITLE
Could the plugin-template.c example be extended please

### DISCUSSION THREAD

**Discussion by logos-maker** [free-audio/clap#120] [Ideas]
I would like to write a CLAP plug in pure C code. And I compiled the plugin-template and tried it in bitwig, and it works.

Is there a more complete example in C code, than the plugin-template? It would be most helpful. In fact anything more would be helpful.

**Comment by baconpaul** [comment]
we have a complete example in C++ in the `clap-saw-demo` and the `clap-plugins` projects but we don't have a complete demo in pure c.

If you choose to skip the GUI, the `clap-saw-demo` should be pretty easy to port to pure c though, if that's of interest to you?

definitely the case that most plugin cases and all the public plugin examples are using the C++ helper wrappers now so we are a bit short on documentation / examples on this front.

**Reply by logos-maker** [reply]
The most important thing missing in plugin-template.c is the usage of the params extension.

**Comment by debrisapron** [comment]
This is actually one of the most interesting things about CLAP from my point of view. I really don't like C++ and am excited for the chance to make audio plugins in a simpler language, but the current example is too bare-bones for me to get started on anything meaningful. I would particularly love to see an example integrating with a simple C audio library such as e.g. Miniaudio.

**Reply by baconpaul** [reply]
well i'd be happy to collaborate with folks on porting a version of the clap saw demo to pure C, at least the non-UI part of it. Or at least a basic no-UI C gain or distortion plugin.

the UI in pure C would be a heavy lift considering the frameworks available. but the basics of voice management and params should be doable for the demo without much trouble.

**Reply by debrisapron** [reply]
I'm happy with generic UI for a lot of things, but yeah at some point I'll want to make a UI. Gonna take a look at the options - I wonder if there are usable C bindings for Qt?

Anyway I'm gonna have a look at your distortion example this weekend & see how I go.

**Reply by baconpaul** [reply]
Yeah cross platform C++ frameworks suitable for plugins are few and far between. JUCE is closest but that's C++. VSTGUI is also an option, but again, is C++. We had a long discussion about this in the audio programmer discord in the c++ channel this week, actually, since we've been wondering what to use for simple examples.

Limit yourself to C and not C++ and I think you may be just coding to the native API on windows and, well, basically assuming objective-c is close enough on mac. And linux would be hard going.  Happy to hear what you conclude!

**Comment by baconpaul** [comment]
Anyway

https://github.com/baconpaul/clap-c99-distortion

there's a distortion plugin

I want to make some changes to it

1: ASYM mode sucks and i wanna make it a wave folder and
2: Want to implement state

but will finish it up this week. It does params already.

Hope that helps.

**Comment by baconpaul** [comment]
OK I added to that c99 distortion full param and state support and pushed it to baconpaul

I'll add some comments and then migrate it to surge-synth-team sometime soon enough but it should show you what you need.

The DSP is painfully naive. You almost definitely don't want to use it musically.

**Reply by baconpaul** [reply]
Oh and I welcome PRs and improvements of course.
